_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+{% highlight Rust %}
+let mut encoder = self.channel.receiver.recv().unwrap(); // get the next command buffer. Error processing is cut off for clarity here
+encoder.clear(&self.out_color, [0.0, 0.0, 0.0, 1.0]);
+encoder.update_constant_buffer(...); // called multiple times
+encoder.draw(...); // called multiple times
+self.channel.sender.send(encoder); // send the command buffer for submission
+{% endhighlight %}
+
+
+## Resource Views
+
+In OpenGL a texture actually represents multiple things:
+  - the allocated video memory storage for the texel data
+  - the way it's being accessed (format, binding)
+  - the sampler state to be used with it
+
+Dx11 introduced a conceptually new look at the graphics objects. It separated all these three concepts into different things. First, you create a texture object, which you can only have limited things to do with. If you want to use it as a render target, you create a render target view (RTV) for it. For sampling, you create a shader resource view (SRV), and may be an extra sampler object. There are also depth-stencil views (DSV) and unordered/ordered access views (UAV, OAV). You can have multiple different views into the same resource, as long as they are compatible (can't see a single-channel 16 bit format as RGBA8, for example).
+
+In GFX, when you create a texture you only provide the surface format, which dictates how many bits each channel has. Then, for the view creation, you specify how these bits are interpret (normalized integers, floats, raw ints, etc). Note: the resources are created by a `gfx::Factory` object, which may or may not be sendable/clonable, depending on the backend features. Example:
+{% highlight Rust %}
+let texture = try!(factory.create_texture(kind, levels, SHADER_RESOURCE | RENDER_TARGET, Usage::GpuOnly, None)); // providing the storage configuration and requesting the capabilities
+let resource = try!(factory.view_texture_as_shader_resource::<Srgba8>(&texture, (0, levels-1), format::Swizzle::new())); // creating an SRV for the whole mipmap range
+let target = try!(factory.view_texture_as_render_target(&texture, 0, None)); // creating an RTV for level 0
+{% endhighlight %}
+
+There are helper methods to create the texture with the views for your convenience (`create_render_target`, `create_depth_stencil`, etc). Later, the texture object itself does not participate in any rendering. It can only be used for updating the contents and reading them back. Typically, you'd ignore this object and only work with its views, given that they hold the parent object automatically.
+
+Note: applied to OpenGL, the GFX texture and RTV/DSV are mapping to the same thing (GL texture). The GL render buffer is used instead if `SHADER_RESOURCE` capability is not requested.
```

Comment:
I'll reword this. `SHADER_RESROUCE` is the capability of a texture object, not GFX context, so I'll try to avoid the confusion.